File: src/tools/read-file.ts
import { z } from "zod";
import type { GitLabClient } from "../gitlab/client.js";
import { addLineNumbers, truncateLine, isBinaryContent, decodeFileContent, formatSize } from "../utils/format.js";

export const readFileSchema = z.object({
  project_id: z.string().describe("GitLab project ID or path (e.g. 'mygroup/myproject')"),
  file_path: z.string().describe("Path to the file in the repository"),
  ref: z.string().optional().describe("Branch, tag, or commit SHA. Defaults to the project's default branch"),
  start_line: z.number().optional().describe("Start reading from this line number (1-based). Default: 1"),
  end_line: z.number().optional().describe("Stop reading at this line number (inclusive). Default: end of file"),
  max_lines: z.number().optional().describe("Maximum number of lines to return. Default: 500"),
});

export type ReadFileInput = z.infer<typeof readFileSchema>;

export async function readFile(client: GitLabClient, input: ReadFileInput): Promise<string> {
  const ref = input.ref ?? (await client.getDefaultBranch(input.project_id));
  const maxLines = input.max_lines ?? 500;

  const file = await client.getFileContent(input.project_id, input.file_path, ref);

  // Binary detection
  if (isBinaryContent(file.content)) {
    return `Binary file: ${input.file_path} (${formatSize(file.size)})\nCannot display binary content. Use GitLab UI to view this file.`;
  }

  const fullContent = decodeFileContent(file.content);
  const allLines = fullContent.split("\n");
  const totalLines = allLines.length;

  // Apply line range
  const start = Math.max(1, input.start_line ?? 1);
  const end = Math.min(totalLines, input.end_line ?? totalLines);
  let selectedLines = allLines.slice(start - 1, end);

  // Apply max_lines truncation
  let truncated = false;
  if (selectedLines.length > maxLines) {
    selectedLines = selectedLines.slice(0, maxLines);
    truncated = true;
  }

  // Truncate long lines (minified file protection)
  const processed = selectedLines.map((l) => truncateLine(l, 500));

  // Format with line numbers
  const output = addLineNumbers(processed.join("\n"), start);

  // Build header
  const header = `File: ${input.file_path} (${formatSize(file.size)}, ${totalLines} lines, ref: ${ref})`;
  const range = `Showing lines ${start}-${start + selectedLines.length - 1} of ${totalLines}`;

  let result = `${header}\n${range}\n${"─".repeat(60)}\n${output}`;

  if (truncated) {
    result += `\n${"─".repeat(60)}\n⚠️ Output truncated at ${maxLines} lines. Use start_line/end_line to read remaining content.`;
  }

  return result;
}

File: internal/tools/read_file.go
package tools

import (
	"context"
	"fmt"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
	"github.com/nanami7777777/gitlab-code-reader-mcp/internal/gitlab"
)

func ReadFile(client *gitlab.Client) (mcp.Tool, server.ToolHandlerFunc) {
	tool := mcp.NewTool("gl_read_file",
		mcp.WithDescription("Read a file from a GitLab repository with line numbers. Supports line ranges for large files. Like Claude Code's Read tool but for remote repos."),
		mcp.WithString("project_id", mcp.Required(), mcp.Description("GitLab project ID or path (e.g. 'mygroup/myproject')")),
		mcp.WithString("file_path", mcp.Required(), mcp.Description("Path to the file in the repository")),
		mcp.WithString("ref", mcp.Description("Branch, tag, or commit SHA. Defaults to the project's default branch")),
		mcp.WithNumber("start_line", mcp.Description("Start reading from this line number (1-based). Default: 1")),
		mcp.WithNumber("end_line", mcp.Description("Stop reading at this line number (inclusive). Default: end of file")),
		mcp.WithNumber("max_lines", mcp.Description("Maximum number of lines to return. Default: 500")),
	)

	handler := func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		args := getArgs(req.Params.Arguments)
		projectID := argStr(args, "project_id")
		filePath := argStr(args, "file_path")

		ref := argStr(args, "ref")
		if ref == "" {
			r, err := client.DefaultBranch(projectID)
			if err != nil {
				return mcp.NewToolResultError(fmt.Sprintf("❌ Error: %v", err)), nil
			}
			ref = r
		}

		maxLines := argInt(args, "max_lines", "", 500)

		file, err := client.GetFileContent(projectID, filePath, ref)
		if err != nil {
			return mcp.NewToolResultError(fmt.Sprintf("❌ Error: %v", err)), nil
		}

		if isBinary(file.Content) {
			return mcp.NewToolResultText(fmt.Sprintf("Binary file: %s (%s)\nCannot display binary content.", filePath, formatSize(file.Size))), nil
		}

		content, err := decodeBase64(file.Content)
		if err != nil {
			return mcp.NewToolResultError(fmt.Sprintf("❌ Decode error: %v", err)), nil
		}

		allLines := strings.Split(content, "\n")
		totalLines := len(allLines)

		startLine := argInt(args, "start_line", "", 1)
		endLine := argInt(args, "end_line", "", totalLines)

		startLine = maxInt(1, startLine)
		endLine = minInt(totalLines, endLine)
		selected := allLines[startLine-1 : endLine]

		truncated := false
		if len(selected) > maxLines {
			selected = selected[:maxLines]
			truncated = true
		}

		for i, l := range selected {
			selected[i] = truncateLine(l, 500)
		}

		output := addLineNumbers(strings.Join(selected, "\n"), startLine)
		header := fmt.Sprintf("File: %s (%s, %d lines, ref: %s)", filePath, formatSize(file.Size), totalLines, ref)
		rangeInfo := fmt.Sprintf("Showing lines %d-%d of %d", startLine, startLine+len(selected)-1, totalLines)

		result := fmt.Sprintf("%s\n%s\n%s\n%s", header, rangeInfo, strings.Repeat("─", 60), output)
		if truncated {
			result += fmt.Sprintf("\n%s\n⚠️ Output truncated at %d lines. Use start_line/end_line to read remaining content.", strings.Repeat("─", 60), maxLines)
		}
		return mcp.NewToolResultText(result), nil
	}

	return tool, handler
}
